xapian-applications/omega/clickmodel/simplifieddbn.cc

```diff
@@ -0,0 +1,207 @@
+/** @file simplifieddbn.cc
+ * @brief SimplifiedDBN class - the Simplified DBN click model.
+ */
+/* Copyright (C) 2017 Vivek Pal
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+#include <config.h>
+
+#include "simplifieddbn.h"
+
+#include <algorithm>
+#include <iomanip>
+#include <iostream>
+#include <fstream>
+#include <sstream>
+#include <string>
+#include <vector>
+
+using namespace std;
+
+#define QID 0
+#define DOCIDS 1
+#define CLICKS 2
+
+string
+SimplifiedDBN::name()
+{
+    return "SimplifiedDBN";
+}
+
+static vector<string>
+get_docid_list(const vector<vector<string>> &sessions)
+{
+    vector<string> docids;
+    string docid;
+    for (size_t i = 0; i < sessions.size(); ++i) {
+	for (size_t j = 0; j <= sessions[i][DOCIDS].length(); ++j) {
+	    if (sessions[i][DOCIDS][j] != ',' && sessions[i][1][j] != '\0') {
+		docid += sessions[i][DOCIDS][j];
+	    } else {
+		docids.push_back(docid);
+		docid.clear();
+	    }
+	}
+    }
+    return docids;
+}
+
+static vector<int>
+get_click_list(const vector<vector<string>> &sessions)
+{
+    vector<int> clicks;
+    string clickstring, clickcount;
+    int click = 0;
+    for (size_t i = 0; i < sessions.size(); ++i) {
+	for (size_t j = 0; j <= sessions[i][CLICKS].length(); ++j) {
+	    if (sessions[i][CLICKS][j] == '\0') {
+		clicks.push_back(click);
+	    } else if (sessions[i][CLICKS][j] != ',') {
+		// Get clickstring of the form "docid:click_count".
+		clickstring += sessions[i][CLICKS][j];
+
+		// Get click count string.
+		int delimiter_pos = clickstring.find(':');
+		clickcount = clickstring.substr(delimiter_pos + 1,
+			     clickstring.length() - 1 - delimiter_pos);
+
+		// Convert click count string to an integer.
+		stringstream ss(clickcount);
+		ss >> click;
```

Comment:
This splits and parses the number in `clickstring` for every character appended to it.

I think you want to merge the `,` and `\0` cases into one, and split and parse the number there instead.  Essentially the same loop structure that you have in `get_docid_list()` above.